internal/locate/region_cache.go

```diff
@@ -726,9 +726,48 @@ func NewRegionCache(pdClient pd.Client, opt ...RegionCacheOpt) *RegionCache {
 		// cache GC is incompatible with cache refresh
 		c.bg.schedule(c.gcRoundFunc(cleanRegionNumPerRound), cleanCacheInterval)
 	}
+	c.bg.schedule(
+		func(ctx context.Context, _ time.Time) bool {
+			refreshFullStoreList(ctx, c.PDClient(), c.stores)
+			return false
+		}, refreshStoreListInterval,
+	)
 	return c
 }
 
+// Try to refresh full store list. Errors are ignored.
+func refreshFullStoreList(ctx context.Context, pdClient pd.Client, stores storeCache) {
+	storeList, err := pdClient.GetAllStores(ctx)
+	if err != nil {
+		logutil.Logger(ctx).Info("refresh full store list failed", zap.Error(err))
+		return
+	}
+	for _, store := range storeList {
+		_, exist := stores.get(store.GetId())
+		if exist {
+			continue
+		}
+		s := stores.getOrInsertDefault(store.GetId())
+		// GetAllStores is supposed to return only Up and Offline stores.
+		// This check is being defensive and to make it consistent with store resolve code.
+		if store == nil || store.GetState() == metapb.StoreState_Tombstone {
+			s.setResolveState(tombstone)
+			continue
+		}
+		addr := store.GetAddress()
+		if addr == "" {
+			continue
+		}
```

Comment:
Can we move this before `stores.getOrInsertDefault`?  `stores.getOrInsertDefault(store.GetId())` will insert a store{id:0} if `store == nil`, and inserting a tombstone or a store whose addr is unknown seems pointless.